namecache/namecache.go

```diff
@@ -0,0 +1,258 @@
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	if _, ok := nc.follows[name]; ok {
+		return fmt.Errorf("Already following %s", name)
+	}
+
+	ctx, cancel := context.WithCancel(nc.ctx)
+	go nc.followName(ctx, name, pinit, followInterval)
+	nc.follows[name] = cancel
+
+	return nil
+}
+
+// Unfollow cancels a follow
+func (nc *nameCache) Unfollow(name string) error {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	cancel, ok := nc.follows[name]
+	if !ok {
+		return fmt.Errorf("Unknown name %s", name)
+	}
+
+	cancel()
+	delete(nc.follows, name)
+	return nil
+}
+
+// ListFollows returns a list of names currently being followed
+func (nc *nameCache) ListFollows() []string {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	follows := make([]string, 0, len(nc.follows))
+	for name, _ := range nc.follows {
+		follows = append(follows, name)
+	}
+
+	return follows
+}
+
+func (nc *nameCache) followName(ctx context.Context, name string, pinit bool, followInterval time.Duration) {
+	// if cid != nil, we have created a new pin that is updated on changes and
+	// unpinned on cancel
+	cid, err := nc.resolveAndPin(ctx, name, pinit)
+	if err != nil {
+		log.Errorf("Error following %s: %s", name, err.Error())
+	}
+
+	ticker := time.NewTicker(followInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if cid != nil {
+				cid, err = nc.resolveAndUpdate(ctx, name, cid)
+			} else {
+				cid, err = nc.resolveAndPin(ctx, name, pinit)
+			}
+
+			if err != nil {
+				log.Errorf("Error following %s: %s", name, err.Error())
+			}
+
+		case <-ctx.Done():
+			if cid != nil {
+				err = nc.unpin(cid)
+				if err != nil {
+					log.Errorf("Error unpinning %s: %s", name, err.Error())
+				}
+			}
+			return
+		}
+	}
+}
+
+func (nc *nameCache) resolveAndPin(ctx context.Context, name string, pinit bool) (*cid.Cid, error) {
+	ptr, err := nc.resolve(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !pinit {
+		return nil, nil
+	}
+
+	cid, err := pathToCid(ptr)
+	if err != nil {
+		return nil, err
+	}
+
+	defer nc.bstore.PinLock().Unlock()
+
+	_, pinned, err := nc.pinning.IsPinned(cid)
+	if pinned || err != nil {
+		return nil, err
+	}
+
+	n, err := nc.pathToNode(ctx, ptr)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("pinning %s", cid.String())
+
+	err = nc.pinning.Pin(ctx, n, true)
+	if err != nil {
+		return nil, err
+	}
+
+	err = nc.pinning.Flush()
+
+	return cid, err
+}
+
+func (nc *nameCache) resolveAndUpdate(ctx context.Context, name string, cid *cid.Cid) (*cid.Cid, error) {
+
+	ptr, err := nc.resolve(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	ncid, err := pathToCid(ptr)
+	if err != nil {
+		return nil, err
+	}
+
+	if ncid.Equals(cid) {
+		return cid, nil
+	}
+
+	defer nc.bstore.PinLock().Unlock()
+
+	log.Debugf("Updating pin %s -> %s", cid.String(), ncid.String())
```

Comment:
Should also check `nc.pinning.IsPinned(ncid)` as is done above and only unpin the old cid if the new is present.

Reply:
I am not sure I understand why you want this check here. I do a pin update, no explicit unpin.